pkg/upstream/rewrite.go

```diff
@@ -0,0 +1,78 @@
+package upstream
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/justinas/alice"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/middleware"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/app/pagewriter"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+)
+
+// newRewritePath creates a new middleware that will rewrite the request URI
+// path before handing the request to the next server.
+func newRewritePath(rewriteRegExp *regexp.Regexp, rewriteTarget string, writer pagewriter.Writer) alice.Constructor {
+	return func(next http.Handler) http.Handler {
+		return rewritePath(rewriteRegExp, rewriteTarget, writer, next)
+	}
+}
+
+// rewritePath uses the regexp to rewrite the request URI based on the provided
+// rewriteTarget.
+func rewritePath(rewriteRegExp *regexp.Regexp, rewriteTarget string, writer pagewriter.Writer, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		reqURL, err := url.ParseRequestURI(req.RequestURI)
+		if err != nil {
+			logger.Errorf("could not parse request URI: %v", err)
+			writer.WriteErrorPage(rw, pagewriter.ErrorPageOpts{
+				Status:    http.StatusInternalServerError,
+				RequestID: middleware.GetRequestScope(req).RequestID,
+				AppError:  fmt.Sprintf("Could not parse request URI: %v", err),
+			})
+			return
+		}
+
+		// Use the regex to rewrite the request path before proxying to the upstream.
+		newURI := rewriteRegExp.ReplaceAllString(reqURL.Path, rewriteTarget)
+		reqURL.Path, reqURL.RawQuery, err = splitPathAndQuery(reqURL.Query(), newURI)
+		if err != nil {
+			logger.Errorf("could not parse rewrite URI: %v", err)
+			writer.WriteErrorPage(rw, pagewriter.ErrorPageOpts{
+				Status:    http.StatusInternalServerError,
+				RequestID: middleware.GetRequestScope(req).RequestID,
+				AppError:  fmt.Sprintf("Could not parse rewrite URI: %v", err),
+			})
+			return
+		}
+
+		req.RequestURI = reqURL.String()
+		next.ServeHTTP(rw, req)
+	})
+}
+
+// splitPathAndQuery splits the rewritten path into the URL Path and the URL
+// raw query. Any rewritten query values are appended to the original query
```

Comment:
What happens with duplicates? Which takes priority?  Can you add a test case for this scenario?

I can see this as a future attack vector if we have any code regressions where somehow querystrings are used to get parameters into the backend that would normally be internal to the rewrite logic here.

Reply:
I've added a comment here to expand on this a little

All values, from the original and the rewrite will be handled and they will be added in order, original and then rewrite, as per the URL encoding library does it.

> I can see this as a future attack vector if we have any code regressions where somehow querystrings are used to get parameters into the backend that would normally be internal to the rewrite logic here.

I'm not sure I'm following, the point of this is to allow users to explicitly make sure all requests have a particular query string value, eg they could ensure all upstream requests have `?auth=OAuth2-Proxy` as a parameter if they so desired

Reply:
> I'm not sure I'm following, the point of this is to allow users to explicitly make sure all requests have a particular query string value, eg they could ensure all upstream requests have `?auth=OAuth2-Proxy` as a parameter if they so desired

The scenario I was thinking of is if I as a client added `?auth=EVILOAuth2-Proxy` (dumb example of something that changes backend logic, I dunno 😅 ). So now we get `?auth=EVILOAuth2-Proxy&auth=OAuth2-Proxy`.

I think the root of this thought process if should these server side added querystrings REPLACE rather than DUPLICATE any that happened to also be in the client request?

Reply:
At the moment, any user added query string would be proxied to the backend anyway, I don't think we want to be stripping anything here (unless someone really comes at us with a valid use case), the only thought I had on query string literally was what would happen if someone put one in their rewrite target

I think for now we should merge as is and move on, if a customer actually ends up using query strings like that, i think it's probably up to them to work out what they do and don't trust